Issue #1485: Simple extension sorting system

dconrad5 opened the pull request:
Explicit ordering of extensions which are run on_episode_downloaded.

Thoughts on the order? Any problems with fixing the issue this way?

auouymous commented:
I assume this ordering would work for most cases. It wouldn't work if the user needed a command to run early, and would break their workflow if it currently does that for them. That is easily fixed by copying or renaming the extension with a lower priority. And I assume command_on_download is the only extension anyone would want to change the order of, or have multiple instances of it.

The biggest problem with this PR is that it breaks existing installs because gpodder stores the enabled extensions without the prefix and users would need to manually enable all of these extensions. It will probably also break settings since the extension's settings are stored under its filename.

It would be better if gpodder stripped the prefix when storing in settings, and scanned for a prefixed file matching. It should probably fail if multiple files are found with different prefixes, such as `15_command_on_download.py` and `90_command_on_download.py`. This way the user can change prefixes and the extensions just work. If you wanted to run multiple commands, with same prefix or different prefixes, you would need to give each copy of the extension a unique name: `15_command_on_download_1.py`, `90_command_on_download.py` and `90_command_on_download_2.py`. There would then be three separate extensions enabled with settings called: `command_on_download_1`, `command_on_download` and `command_on_download_2`.

Also, we need to make sure the directory listing is actually sorted and this isn't just working because the files are in that order in the directory table. I don't know if Python guarantees sorted order, but if not, gpodder must sort them. Gpodder does load from multiple directories, so the sorting must happen across them, as a whole. In the above example, the suffixed copies of command_on_download would be in the user's Extension directory, with the original in `/usr/share`.

dconrad5 commented:
Thanks for the thoughts. I added 7dccd47, to strip the prefix, if found, from the filename when it stores the extension name field. I think some concerns are largely already taken care of, fortunately. (and I really need to remember to do a "git pull --merge" instead of "git pull --rebase" when pulling in upstream changes to avoid commit references changing...)

> I assume this ordering would work for most cases. It wouldn't work if the user needed a command to run early, and would break their workflow if it currently does that for them. That is easily fixed by copying or renaming the extension with a lower priority. And I assume command_on_download is the only extension anyone would want to change the order of, or have multiple instances of it.
> 
> The biggest problem with this PR is that it breaks existing installs because gpodder stores the enabled extensions without the prefix and users would need to manually enable all of these extensions. It will probably also break settings since the extension's settings are stored under its filename.

As far as I can tell the settings are already stored under the extension name rather than the filename, and always referenced by the extension name. The filename is only ever used to actually refer to the file location itself. 

From my testing this seems to be the case in practice as well. My test is as follows:

1. enable audio_converter and command_on_download - at this point they're "20_audio_converter.py" and "90_command_on_download.py"
2. Customize the default command_on_download zenity command in some way (I put "custom" at the beginning of the string it prints). Note that the config setting is still "extension.command_on_download.command", but the filename is "90_command_on_download.py".
3. Enable the "use_ogg" option for audio_converter
4. Download an mp3 file - note that the file is converted prior to displaying the zenity message, and that it both reflects the new filename with ogg extension and that the string customization is present
5. Quit gpodder, rename 90_command_on_download.py to 15_command_on_download.py, and restart gpodder
6. run step 4 again. Note that the zenity message still refers to the filename with an mp3 extension and that it still contains the string customization. Also note that the file is not converted to ogg until the zenity message is dismissed.
7. The same test can be run in a different way by renaming command_on_download back to 90_command_on_download.py and moving audio_converter to 91_audio_converter.py. Note that the use_ogg setting is retained in this test as well.

According to the above test, I would say it's fundamentally working as intended.

> It would be better if gpodder stripped the prefix when storing in settings, and scanned for a prefixed file matching. It should probably fail if multiple files are found with different prefixes, such as `15_command_on_download.py` and `90_command_on_download.py`. This way the user can change prefixes and the extensions just work. If you wanted to run multiple commands, with same prefix or different prefixes, you would need to give each copy of the extension a unique name: `15_command_on_download_1.py`, `90_command_on_download.py` and `90_command_on_download_2.py`. There would then be three separate extensions enabled with settings called: `command_on_download_1`, `command_on_download` and `command_on_download_2`.

I think as is right now, multiple extensions with the same "base name" and different prefixes will probably fight over the config settings, possibly in somewhat unpredictable ways. It would probably be a reasonable improvement to handle this in a predictable way, agreed.

But the base functionality seems to be there already - I can rename an extension file and it still finds the already-present settings for the base name.

> Also, we need to make sure the directory listing is actually sorted and this isn't just working because the files are in that order in the directory table. I don't know if Python guarantees sorted order, but if not, gpodder must sort them. Gpodder does load from multiple directories, so the sorting must happen across them, as a whole. In the above example, the suffixed copies of command_on_download would be in the user's Extension directory, with the original in `/usr/share`.

Looks like we sort the order ourselves here:
https://github.com/gpodder/gpodder/blob/7dccd47e9cb7a790f83cd611f8ae45ee2d3dec51/src/gpodder/extensions.py#L396
Looks like this is already done after searching all relevant folders and gathering everything up. I changed it so that it sorts by the filename rather than the extension name, which was the prior behavior.

dconrad5 commented:
> > It would be better if gpodder stripped the prefix when storing in settings, and scanned for a prefixed file matching. It should probably fail if multiple files are found with different prefixes, such as `15_command_on_download.py` and `90_command_on_download.py`. This way the user can change prefixes and the extensions just work. If you wanted to run multiple commands, with same prefix or different prefixes, you would need to give each copy of the extension a unique name: `15_command_on_download_1.py`, `90_command_on_download.py` and `90_command_on_download_2.py`. There would then be three separate extensions enabled with settings called: `command_on_download_1`, `command_on_download` and `command_on_download_2`.
> 
> I think as is right now, multiple extensions with the same "base name" and different prefixes will probably fight over the config settings, possibly in somewhat unpredictable ways. It would probably be a reasonable improvement to handle this in a predictable way, agreed.

Upon a second look, gpodder already intentionally overrides extensions found with matching names in order to allow extensions in the user directory to take precedence. So I believe this is basically already taken care of. I added a logger info message to make it easier to tell that this is happening.

Relevant:

https://github.com/gpodder/gpodder/blob/faf110b45d0afd9334f6bb9bca9d74a1374d491d/src/gpodder/extensions.py#L374-L403

dconrad5 commented:
Upon third thought, I believe the filename field also contains the entire path to the extension, so we might get unexpected results when also loading from the user directory... I think I may have to rework the sorting to explicitly find the prefix somehow.

dconrad5 commented:
> Upon third thought, I believe the filename field also contains the entire path to the extension, so we might get unexpected results when also loading from the user directory... I think I may have to rework the sorting to explicitly find the prefix somehow.

Added commit a128aa2, which explicitly finds the priority independent of the extensions path. Seems to work well from my testing, including using the user's extension folder to override system extensions.

The extension class now has an "extpriority" variable which might come in handy in the future.

dconrad5 commented:
Ok, I think this is ready for review, I can't think of any more gotchas at the moment.

elelay commented on src/gpodder/extensions.py:
```diff
@@ -380,16 +391,31 @@ def _find_extensions(self):
-            extensions[name] = filename
 
-        return sorted(extensions.items())
+            # get ordering prefix
+            priority = self._get_prefix(name)
+
+            # strip ordering prefix from name, if present
+            name = re.sub(r'^[0-9]*_', '', name)
+            try:
+                if extensions[name] is not None:
+                    logger.info("extension at %s will be ignored in favor of %s", extensions[name][1], filename)
+            except:
+                None
+            extensions[name] = (priority, filename)
```
Untested, but maybe you could get rid of _get_prefix() and write it like that.
I prefer playing with the prefix in a single place. I also prefer to use `key in dict` to avoid the `try/except`

```suggestion
            priority = 99
            m = re.fullmatch(r'^([0-9]*)_(.+)', name)
            if m:
              if m.group(1):
                # get ordering prefix
                priority = int(m.group(1))
              # strip ordering prefix from name (or leading _)
              name = m.group(2)
              if name in extensions:
                    logger.info("extension at %s will be ignored in favor of %s", extensions[name][1], filename)
            extensions[name] = (priority, filename)
```

dconrad5 replied:
This seems like a bit of pythonism I'm not familiar with, but I'll parse through it and see if it does the same thing, sure.

dconrad5 replied:
After parsing through it and trying it out, yeah it seems reasonable enough. I originally thought there would be other places _get_prefix() would be used, but that was before we were saving it as its own value.

elelay reviewed (changes requested):
ok in principle, but please have a look at the suggestions (open to discussion)

dconrad5 commented on src/gpodder/extensions.py:
```diff
@@ -380,16 +391,31 @@ def _find_extensions(self):
+
+        # sort by priority - extensions with same priority will retain the order
+        # in which they were found.
+        return sorted(extensions.items(), key=lambda i: i[1][0])
```
That's a good question. I think right now they will tend to be, but the behavior right now isn't well defined I don't think. I'll have to play with it, maybe we can detect that situation and do something predictable.

dconrad5 replied:
Looks like that works fine, committing.

auouymous commented on src/gpodder/extensions.py:
```diff
@@ -333,10 +334,10 @@ def __init__(self, core):
-        for name, filename in self._find_extensions():
-            logger.debug('Found extension "%s" in %s', name, filename)
+        for name, (priority, filename) in self._find_extensions():
+            logger.debug('Found extension "%s", priority %s, in %s', name, str(priority), filename)
```
The `%s` would also need to be changed to `%d`, since priority is an integer.

elelay replied:
hum, no need to. It's not C. `%s` will result in `str()` be called on the argument if it's not a string already.
```
Python 3.13.1 (main, Dec  4 2024, 18:05:56) [GCC 14.2.1 20240910] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> "%s" % 1
'1'
```

dconrad5 replied:
Looks fine, committed.

auouymous commented:
The priority fails when you have a duplicate extension at different priorities. I added test.py, followed by 66_test.py and it correctly used 66. I added 80_test.py in share and it used 66. But I then symlinked 42_test.py to 80_test.py, and it still used 66.

The separation of share and user extensions guarantees the user can override shared extensions. However, same name extensions in the user directory (or shared directory) are randomly sorted and priority is ignored. Sorting them in descending order fixes that issue.

```python
           self.filenames = sorted(glob.glob(builtins), reverse=True) \
                            + sorted(glob.glob(user_extensions), reverse=True)
```

> logger.info("extension at %s will be ignored in favor of %s", extensions[name][1], filename)

The following switches to a warning, which is always logged and written to console, when an extension is duplicated in share or duplicated in user, but not when user overrides share. This will help catch errors when a user duplicates thinking it they can run the extension twice, forgetting to add a suffix.

```python
            if name in extensions:
                if filename.startswith(gpodder.home) == extensions[name][1].startswith(gpodder.home):
                    logger.warning("extension at %s will be ignored in favor of %s", extensions[name][1], filename)
                else:
                    logger.info("extension at %s will be ignored in favor of %s", extensions[name][1], filename)
```

> The extension class now has an "extpriority" variable which might come in handy in the future.

Why not just "priority"?

dconrad5 commented:
> The priority fails when you have a duplicate extension at different priorities. I added test.py, followed by 66_test.py and it correctly used 66. I added 80_test.py in share and it used 66. But I then symlinked 42_test.py to 80_test.py, and it still used 66.
> 
> The separation of share and user extensions guarantees the user can override shared extensions. However, same name extensions in the user directory (or shared directory) are randomly sorted and priority is ignored. Sorting them in descending order fixes that issue.
> 
> ```python
>            self.filenames = sorted(glob.glob(builtins), reverse=True) \
>                             + sorted(glob.glob(user_extensions), reverse=True)
> ```
> 

I'll have to take a look at this, but seems reasonable enough.

> > logger.info("extension at %s will be ignored in favor of %s", extensions[name][1], filename)
> 
> The following switches to a warning, which is always logged and written to console, when an extension is duplicated in share or duplicated in user, but not when user overrides share. This will help catch errors when a user duplicates thinking it they can run the extension twice, forgetting to add a suffix.
> 
> ```python
>             if name in extensions:
>                 if filename.startswith(gpodder.home) == extensions[name][1].startswith(gpodder.home):
>                     logger.warning("extension at %s will be ignored in favor of %s", extensions[name][1], filename)
>                 else:
>                     logger.info("extension at %s will be ignored in favor of %s", extensions[name][1], filename)
> ```
> 

I'll add this, seems reasonable enough.

> > The extension class now has an "extpriority" variable which might come in handy in the future.
> 
> Why not just "priority"?

No real reason, just the name I chose. "priority" works fine enough for me.

auouymous commented:
> > The priority fails when you have a duplicate extension at different priorities. I added test.py, followed by 66_test.py and it correctly used 66. I added 80_test.py in share and it used 66. But I then symlinked 42_test.py to 80_test.py, and it still used 66.
> 
> I'm not entirely sure I follow the issue. Is the symlink (42_test.py) in share, or user directory? If it's in share, I think the intended behavior is for it to continue using what it finds in user directory, regardless of whether the version in user directory or share has a lower prefix. Or I might be misunderstanding?

The 42 was in user. The `glob.glob()` calls produce an unsorted listing of files, or they could be sorted but in ascending order, so 42 is (can be) seen before 66, and the code discards all but the last, no matter what the priority is.


> > The separation of share and user extensions guarantees the user can override shared extensions. However, same name extensions in the user directory (or shared directory) are randomly sorted and priority is ignored. Sorting them in descending order fixes that issue.
> > ```python
> >            self.filenames = sorted(glob.glob(builtins), reverse=True) \
> >                             + sorted(glob.glob(user_extensions), reverse=True)
> > ```
> 
> I'm not entirely sure I follow this either (or is this referring to the same thing as above?) - don't we sort at the end, after all the extension names and priorities have been parsed?

This sorts the input files in descending order so you get [test, 66_test, 42_test] instead of [42_test, 66_test, test] or some random ordering. Any file without a priority gets sorted first, which is good because they are assumed to be 99. Now when it discards all but the last, it will be keeping 42, which is the lowest priority.

The final sorting at return is simply sorting the list of selected extension files for loading purposes.

auouymous commented:
I forgot this PR touched a lot of extension files and merged 552db50b8e4c6dac663fb8409e295a7a4be5f01c. It caused some conflicts with the renamed files, sorry about that.

dconrad5 commented:
> > > The priority fails when you have a duplicate extension at different priorities. I added test.py, followed by 66_test.py and it correctly used 66. I added 80_test.py in share and it used 66. But I then symlinked 42_test.py to 80_test.py, and it still used 66.
> > 
> > 
> > I'm not entirely sure I follow the issue. Is the symlink (42_test.py) in share, or user directory? If it's in share, I think the intended behavior is for it to continue using what it finds in user directory, regardless of whether the version in user directory or share has a lower prefix. Or I might be misunderstanding?
> 
> The 42 was in user. The `glob.glob()` calls produce an unsorted listing of files, or they could be sorted but in ascending order, so 42 is (can be) seen before 66, and the code discards all but the last, no matter what the priority is.
> 
> > > The separation of share and user extensions guarantees the user can override shared extensions. However, same name extensions in the user directory (or shared directory) are randomly sorted and priority is ignored. Sorting them in descending order fixes that issue.
> > > ```python
> > >            self.filenames = sorted(glob.glob(builtins), reverse=True) \
> > >                             + sorted(glob.glob(user_extensions), reverse=True)
> > > ```
> > 
> > 
> > I'm not entirely sure I follow this either (or is this referring to the same thing as above?) - don't we sort at the end, after all the extension names and priorities have been parsed?
> 
> This sorts the input files in descending order so you get [test, 66_test, 42_test] instead of [42_test, 66_test, test] or some random ordering. Any file without a priority gets sorted first, which is good because they are assumed to be 99. Now when it discards all but the last, it will be keeping 42, which is the lowest priority.
> 
> The final sorting at return is simply sorting the list of selected extension files for loading purposes.

Ok, I think I'm with you now, and that seems like a reasonable solution - I'll get it added.

dconrad5 commented:
> This sorts the input files in descending order so you get [test, 66_test, 42_test] instead of [42_test, 66_test, test] or some random ordering. Any file without a priority gets sorted first, which is good because they are assumed to be 99. Now when it discards all but the last, it will be keeping 42, which is the lowest priority.
> 
> The final sorting at return is simply sorting the list of selected extension files for loading purposes.

Added as https://github.com/gpodder/gpodder/pull/1698/commits/9a4c3a69d696ad040c567212b3ef3b845d2d49d1, please take a look.

> I forgot this PR touched a lot of extension files and merged [552db50](https://github.com/gpodder/gpodder/commit/552db50b8e4c6dac663fb8409e295a7a4be5f01c). It caused some conflicts with the renamed files, sorry about that.

No worries, got master branch merged into this one, should be mergeable now.

auouymous commented on src/gpodder/extensions.py:
```diff
@@ -334,10 +334,10 @@ def __init__(self, core):
+            logger.debug('Found extension "%s", priority %s, in %s', name, extension_priority, filename)
             config = getattr(core.config.extensions, name)
-            container = ExtensionContainer(self, name, config, filename, extpriority=priority)
+            container = ExtensionContainer(self, name, config, filename, priority=extension_priority)
```
Does it not allow `priority=priority` to be passed to ExtensionContainer?

dconrad5 replied:
Probably, that was just a personal choice that I wanted to have different names to avoid any confusion. I can change it if you would like?

auouymous reviewed (changes requested):
LGTM.

@elelay Are you ready to merge?

thp commented:
Sorry I'm a bit late to the discussion (had the [corresponding documentation change](https://github.com/gpodder/gpodder.github.io/pull/62) in my e-mail notifications).

Would it make sense to explicitly list the relative order of the extension but not an absolute numeric ordering?

systemd calls this [implicit and explicit dependencies](https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html#Automatic%20dependencies). In particular, [`After=` and `Before=`](https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html#Before=) are used to enforce certain orderings without requiring the user to specify an absolute ordering.

This way, no file renaming needs to be in place, and we just need to define (inside the extension) when they would need to be loaded relative to some other extension.

We already have things like `__only_for__` and `__disable_in__`, and although those are all strings, I'd say a tuple-of-strings might be good for those, e.g.:

```
__after__ = ("on_episode_downloaded",)
```

Note that this is a single-element tuple, hence the "," at the end.

We probably don't really need `__before__` and whether we call it `__after__` or `__requires__` (for example) is anyone's guess. Also, if we call it `__requires__`, this could also implicitly mean that we drag in the other extension and enable it (e.g. if an extension needs to be loaded after `on_episode_downloaded`, it might also mean that we need to load `on_episode_downloaded` for the depending extension to work.

thp commented:
And even when not going with `__after__` / `__requires__`, then encoding the numeric sort order inside the file as `__priority__ = 99` instead of in the filename would still be better than renaming files and having people "remember" how to transform the filename in the head to the extension name IMHO (even if it's just "remove the numeric prefix", it's still more mental load than not having to do any transformation).

thp commented:
And as for the (not sure how relevant?) case where the user wants to enforce some additional ordering that isn't needed by everybody, one way to do this would be to create an empty extension in the user directory, and there `__before__` could then also be needed. For example, if I wanted to load the `ted_subtitles.py` extension before `mpris-listener` for whatever reason, I could create an extension with:

```
__after__ = ("ted_subtitles",)
__before__ = ("mpris-listener",)
```

When calculating the extension load order, this would make sure that the user's customization requirement is taken into account. I would only allow this kind of ability to customize if there's a real need for it, as if you do it like this, you also need to take care of resolving circular dependencies.

Not knowing the full picture, I'd just go for `__after__` and implement sorting this way, not doing anything with custom priorities, and not letting the user override the load order any further for now (until the need arises).

Otherwise you will also potentially get bug reports of users having misconfigured their extension load order and it will be one additional data item that you need to request your users from listing when filing a bug report, increasing the problem search space.

dconrad5 commented:
That certainly could be very powerful - I am presuming that this would require a GUI editor in order for the user to change the priorities. I only did this because it was an easy way to achieve something predictable and easy to work with. The method in this PR could certainly be replaced with something much more robust in the future (or this PR could be dropped), but to be honest I don't think I'm going to be the one writing that.

thp commented:
I've had another look at it, and tried to implement it with `__after__`, but the relative ordering in this PR doesn't really make much sense (e.g. `youtube-dl` doesn't need to be loaded before any other extension, as it's not an `on_episode_downloaded()` extension), so I quickly scrapped that idea ;)

However, we have a configuration value that is implicitly sorted: `core.config.extensions.enabled` -- if we just sort the found extensions by the order in which the extensions appear, this allows the user to easily just reorder the list of enabled extensions. No need to rename files, no need to hardcode `__after__` / `__before__` in the extension code.

For now, users would be able to sort the extensions by changing the enabled extensions configuration manually (or disabling all extensions and enabling them in the order they want them to be loaded). In the future, the extensions list UI could allow drag'n'drop of extensions and allow reordering of enabled extensions this way.

Proposed, untested patch:

```patch
diff --git a/src/gpodder/extensions.py b/src/gpodder/extensions.py
index 82d07810..5ca13649 100644
--- a/src/gpodder/extensions.py
+++ b/src/gpodder/extensions.py
@@ -333,7 +333,7 @@ class ExtensionManager(object):
             logger.info('Disabling all extensions (from environment)')
             return
 
-        for name, filename in self._find_extensions():
+        for name, filename in self._find_extensions(enabled_extensions):
             logger.debug('Found extension "%s" in %s', name, filename)
             config = getattr(core.config.extensions, name)
             container = ExtensionContainer(self, name, config, filename)
@@ -371,7 +371,7 @@ class ExtensionManager(object):
                         for x in self.core.config.extensions.enabled
                         if x != container.name]
 
-    def _find_extensions(self):
+    def _find_extensions(self, enabled_extensions):
         extensions = {}
 
         if not self.filenames:
@@ -389,7 +389,18 @@ class ExtensionManager(object):
             name, _ = os.path.splitext(os.path.basename(filename))
             extensions[name] = filename
 
-        return sorted(extensions.items())
+        def extension_sort_key(kv):
+            name, filename = kv
+            try:
+                index = enabled_extensions.index(name)
+            except ValueError:
+                # Any not-enabled extensions will be sorted at the
+                # beginning of the list, mandatory extensions will
+                # be loaded before any other item in the list.
+                index = -1
+            return (index, name)
+
+        return sorted(extensions.items(), key=extension_sort_key)
 
     def get_extensions(self):
         """Get a list of all loaded extensions and their enabled flag."""
```

thp commented on src/gpodder/extensions.py:
```diff
@@ -378,18 +379,41 @@ def _find_extensions(self):
-            extensions[name] = filename
 
-        return sorted(extensions.items())
+            # extensions with no priority get lowest priority, 99
+            priority = 99
+            m = re.fullmatch(r'^([0-9]*)_(.+)', name)
```
> extensions with no priority get lowest priority, 99

..except when I have a file named `100_foo.py`, which has an even lower priority:

```python3
>>> import re
>>> name = "100_foo"
>>> m = re.fullmatch(r'^([0-9]*)_(.+)', name)
>>> bool(m)
True
>>> bool(m.group(1))
True
>>> int(m.group(1))
100
```

Probably would want the regex to be `r'^([0-9]{2})_(.+)'` or something.

thp replied:
Also, `if m.group(1):` could be avoided if the first match group would be `([0-9]+)` and not `([0-9]*)` (but again, would probably need to enforce 2-digit priorities).

dconrad5 replied:
A fair oversight on my part, but I suppose there's no reason 3+ digit priorities wouldn't be valid. Honestly, I could probably just update the comment and documentation, unless you can think of a reason to limit it to 2 digits?

elelay replied:
no need to limit to 2 digits. And you can use `sys.maxint` as default value instead of 99.

I have no strong viewpoint on `([0-9]+)`: how it is currently allows for _extensionname.py to be converted to *extensionname* with the lowest priority. Not sure if it helps...

dconrad5 replied:
Ok, that's easy enough. Before I do that though, would it be worth keeping 99 as the default so that it is possible to have an extention intentionally run last, even after extensions with no priority? I could make it clear in the documentation that ut's possible to make a super-low priority extension as well.

elelay replied:
good point. +1 to keep 99 then.

dconrad5 replied:
> Ok, I'll change the comment and add a note about it in the documentation then.

Done, please see https://github.com/gpodder/gpodder/pull/1698/commits/3a5bca42f05552217918b2ebdd29c0af0bddf19c

dconrad5 replied:
@thp Do you think that is sufficient or would you still like the number of digits limited?

auouymous replied:
I'm good with 100 being lower than the default 99.

thp commented on src/gpodder/extensions.py:
```diff
@@ -378,18 +379,41 @@ def _find_extensions(self):
+                    priority = int(m.group(1))
+                # strip ordering prefix from name (or leading _)
+                name = m.group(2)
+            if name in extensions:
+                if filename.startswith(gpodder.home) == extensions[name][1].startswith(gpodder.home):
+                    logger.warning("extension at %s will be ignored in favor of %s", extensions[name][1], filename)
+                else:
+                    logger.info("extension at %s will be ignored in favor of %s", extensions[name][1], filename)
```
I'd suggest:

- Do query if old extension exists and retrieval of the name in a single command (also avoids `extensions[name][1]`, which one needs to look for what `[1]` actually contains)
- The "warn if in the same directory, info if different directory" logic is better understood with `dirname(...) == dirname(...)` vs `.startswith(home) == .startswith(home)` (also it would cover more cases if we ever source from more than two directories)

```suggestion
            _, previous_filename = extensions.get(name, (None, None))
            if previous_filename is not None:
                if os.path.dirname(gpodder.home) == os.path.dirname(previous_filename):
                    logger.warning("extension at %s will be ignored in favor of %s in the same directory", previous_filename, filename)
                else:
                    logger.info("extension at %s will be ignored in favor of %s", previous_filename, filename)
```

dconrad5 replied:
Took me a minute but I think I see what you're saying, and looks fine to me. Committed.

thp commented on src/gpodder/extensions.py:
```diff
@@ -378,18 +379,41 @@ def _find_extensions(self):
+                    logger.info("extension at %s will be ignored in favor of %s", extensions[name][1], filename)
+            extensions[name] = (priority, filename)
+
+        # sort by priority - extensions with same priority will retain the order
+        # in which they were found.
+        return sorted(extensions.items(), key=lambda i: (i[1][0], i[0]))
```
The comment is actually wrong:

> extensions with same priority will retain the order in which they were found

In fact, we use [`glob.glob()`](https://docs.python.org/3/library/glob.html#glob.glob) when finding the extensions, and the Python docs there say:

> [...] although results are returned in arbitrary order

So it's a good thing they don't retain the order in which they were found. The code did the right thing and sort by name for same-priority items, but the comment was misleading and the code was hard to read (IMHO).

The `key=` function sorts by priority and then by name. This could be made be clearer like this:

```suggestion
        # sort by priority - extensions with same priority will be sorted by name
        def sort_key(kv):
            name, (priority, filename) = kv
            return (priority, name)
            
        return sorted(extensions.items(), key=sort_key)
``````

dconrad5 replied:
The comment used to be correct, I just neglected to update it. I will update it.

Suggestion looks fine by me, whatever works. Seems to work fine from quick testing.

elelay reviewed (commented):
Hi,
I prefer the file numbering prefix instead of the `core.config.extensions.enabled` property because it is materialized in the Extensions directory contents and I'm used to it for some system configuration files.

This PR seems almost ready to merge (pending maybe fixing lowest priority to allow `100_mysuperlowextension.py`).

elelay commented:
Yet, if a user is not happy with our chosen extension order, with `core.config.extensions.enabled` ordering they only have to make a change once, with filename encoding they must do it after each reinstall.

dconrad5 commented:
> Yet, if a user is not happy with our chosen extension order, with `core.config.extensions.enabled` ordering they only have to make a change once, with filename encoding they must do it after each reinstall.

Well, if they do changes in their user extension folder, won't they be able to keep the changes across updates?

elelay commented:
> > Yet, if a user is not happy with our chosen extension order, with `core.config.extensions.enabled` ordering they only have to make a change once, with filename encoding they must do it after each reinstall.
> 
> Well, if they do file numbering changes in their user extension folder, won't they be able to keep the changes across updates?

Let's say they want to run command_on_download before rename_download (the reverse of what we have now). Then they have to rename in the install folder and do it again on upgrade. Or they can copy and rename current version of the extensions to their user extension folder but they have to do it again after upgrade in case the code of these extensions changed.

elelay reviewed (approved):
LGTM

dconrad5 commented:
@thp Are you still wanting to use an alternate method or is this okay with you?

auouymous commented:
> __after__ = ("on_episode_downloaded",)

Filename priority prefix was changed to before/after values inside init system scripts because it is better. And it would also be better for gpodder extensions, however, `command_on_download.py` would benefit from prefixes. It can default to run after other extensions have modified the download, but users can symlink it to an alternate priority to run before another extension, or symlink it to an alternate name and priority to run multiple commands.

> Yet, if a user is not happy with our chosen extension order, with `core.config.extensions.enabled` ordering they only have to make a change once, with filename encoding they must do it after each reinstall.

The user can symlink to shared extensions. Changes to `core.config.extensions.enabled` disappear if you have to disable the extension for whatever reason and then re-enable it. Symlinks are also the safer option because changing `core.config.extensions.enabled` in `Settings.json` can result in a broken config file. I don't know if `edit config` can damage anything other than potentially disabling extensions if you mistype their name.

I don't know if Windows shortcuts can do this since they are not symlinks. This PR does fix the ordering and most users might not need anything other than that, but Windows users wouldn't be able to take advantage of symlinks if shortcuts don't work. Switching to `core.config.extensions.enabled` would require a GUI to easily change ordering, and `__before__` and `__after__` values in extensions to set sane ordering, that way the default ordering just works for most users.

Getting shortcuts to work on Windows might be easier, and it would allow Windows users to run multiple commands without copying  the extension whenever an update changes it.

thp commented:
I‘d still approach it from a point where file renames and symlinks are not
necessary, but if @elelay and @auouymous are fine with it, please just go
ahead :)

dconrad5 ***@***.***> schrieb am Mo. 28. Apr. 2025 um 18:14:

> *dconrad5* left a comment (gpodder/gpodder#1698)
> <https://github.com/gpodder/gpodder/pull/1698#issuecomment-2835783529>
>
> @thp <https://github.com/thp> Are you still wanting to use an alternate
> method or is this okay with you?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gpodder/gpodder/pull/1698#issuecomment-2835783529>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABBASO2WCYQGW7TSNMZCMT23ZHXTAVCNFSM6AAAAABUPHG3S2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMZVG44DGNJSHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

dconrad5 commented:
> I‘d still approach it from a point where file renames and symlinks are not necessary, but if @elelay and @auouymous are fine with it, please just go ahead :) dconrad5 ***@***.***> schrieb am Mo. 28. Apr. 2025 um 18:14:
> […](#)
> *dconrad5* left a comment ([gpodder/gpodder#1698](https://github.com/gpodder/gpodder/pull/1698)) <[#1698 (comment)](https://github.com/gpodder/gpodder/pull/1698#issuecomment-2835783529)> @thp <https://github.com/thp> Are you still wanting to use an alternate method or is this okay with you? — Reply to this email directly, view it on GitHub <[#1698 (comment)](https://github.com/gpodder/gpodder/pull/1698#issuecomment-2835783529)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AABBASO2WCYQGW7TSNMZCMT23ZHXTAVCNFSM6AAAAABUPHG3S2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMZVG44DGNJSHE> . You are receiving this because you were mentioned.Message ID: ***@***.***>

I do appreciate the advice - perhaps a future rework opportunity. I'm glad to get the low-hanging fruit just to have it workable for now.

elelay commented:
> The user can symlink to shared extensions. 

@auouymous you're right: it's totally how it should be done on unix-like OS
